pyquil/quil.py

```diff
@@ -951,7 +984,9 @@ def implicitly_declare_ro(instructions: List[AbstractInstruction]):
     return new_instr
 
 
-def merge_with_pauli_noise(prog_list: Iterable, probabilities: List, qubits: List) -> Program:
+def merge_with_pauli_noise(
+    prog_list: Iterable[Program], probabilities: List[float], qubits: Sequence[Any]
```

Comment:
```suggestion
    prog_list: Iterable[Program], probabilities: List[float], qubits: Sequence[int]
```

I think this is the intention? Probably this would also need to be changed in `define_noisy_gate` and `_create_kraus_pragmas`, etc.

Reply:
6th commit.

Reply:
This breaks `noise.py` because mypy expects a `Iterable[str]`, which is covered by `Sequence[Any]`.

Reply:
I reverted `_create_kraus_pragmas` back to using `Sequence[Any]` in order to pass type checking of `noise.py`.

Reply:
The typechecking error is just mypy being finicky about heterogeneous `List`s.  AFAICT, nothing is actually passing strings as `qubit_indices` to `_create_kraus_pragmas`.  It's just that mypy thinks `List`s should be homogenous (all elements of same type), and in `_create_kraus_pragmas` we have `[name] + list(qubit_indices)` where `name` is declared to be a `str`. Mypy therefore assumes it must be building a list of `str`, which is incompatible with `qubit_indices: Sequence[int]`.

Apparently mypy doesn't make the same assumption about `tuple`s, so you can resolve the error with the following diff:

```diff
diff --git a/pyquil/noise.py b/pyquil/noise.py
index 4e2ce383..6268df03 100644
--- a/pyquil/noise.py
+++ b/pyquil/noise.py
@@ -200,7 +200,7 @@ def _check_kraus_ops(n: int, kraus_ops: Sequence[np.ndarray]) -> None:
 
 
 def _create_kraus_pragmas(
-    name: str, qubit_indices: Sequence[Any], kraus_ops: Sequence[np.ndarray]
+    name: str, qubit_indices: Sequence[int], kraus_ops: Sequence[np.ndarray]
 ) -> List[Pragma]:
     """
     Generate the pragmas to define a Kraus map for a specific gate on some qubits.
@@ -214,7 +214,7 @@ def _create_kraus_pragmas(
     pragmas = [
         Pragma(
             "ADD-KRAUS",
-            [name] + list(qubit_indices),
+            (name,) + tuple(qubit_indices),
             "({})".format(" ".join(map(format_parameter, np.ravel(k)))),
         )
         for k in kraus_ops

```

Reply:
Please make the above change in noise.py as well. Might as well tighten up the type of `qubit_indices` here while we're at it.

Reply:
OK, will be in the 9th commit.